src/urllib3/response.py

```diff
@@ -154,6 +154,20 @@ def _get_decoder(mode):
     return DeflateDecoder()
 
 
+class _HTTPResponseRequest(object):
```

Comment:
Could this be the beginnings of a prebuildable Request class?

Reply:
I, for one, would rather an immutable Request class that has a builder friend. I know this starts to veer more towards Java land with an object that exists to build another, but there are so many foot-guns in generating an HTTP request, I think a builder could provide the necessary guard-rails. At that point, we could have a method on a Request that pre-fills a builder for a redirect. Then since the Request objects are immutable (and ideally comparable) we can say `nextreq = decider.should_retry(...)` and verify `nextreq != req`. I think the latter is important in the event someone tries to stream/upload a file-like object and wants to retry that, we shouldn't let them.

Reply:
I proposed the idea of using a `dataclass` in Discord along with potentially using dataclasses for other objects like Retry and likely the "Trace" object attached to an HTTPResponse. We can make the object immutable to give it the same properties as you mention with `frozen=True`. For Python 3.6 we can require the `dataclasses` module on PyPI or vendor the library until we drop Python 3.6 in v2.

I'm curious about why you mention that we shouldn't allow retrying uploading file like objects? They should be rewindable, can you elaborate on that?

Reply:
@sethmlarson So _most_ file-like objects are rewindable. However, people have been known to do things like try to stream one response into a different request directly because our `HTTPResponse` object behaves like a file. We can't rewind sockets and so a user should be forced to handle that logic themselves. Does that make sense? We can make educated guesses around when it's safe, but I'm pretty sure a unix socket isn't rewindable either, so even if it looks like a local file it might not be perfectly rewindable

Reply:
As for why I'm -1 on dataclasses and +1 on attrs:

- The standard library drifts between versions meaning that we need to support the common denominator in something like this _or_ categorically use the backported PyPI module to get consistency. Specifically, I'm expecting someone to want to steal more of the good parts of `attrs` in 3.10 and beyond and for us to want to use those meaning downstream package distributors whining that we're using the backport version on 3.8 and 3.9 which is a waste of everyone's time

- Attrs is fast and far more battle-tested than it's poor substitute in the stdlib. It has far more features, and will never have the conflict with packagers of "But this is in the stdlib!"

Reply:
@sigmavirus24 I'm in favor of dataclasses over attrs:

* I'm not expecting to use any special features of attrs/dataclasses, mostly need them for the standardized way to mutate a frozen data object, basically if namedtuples had a mutator we'd use them. Also once we migrate to v2 I suspect we'll be dropping Python versions at a more regular cadence now that we've shed Python 2. We won't be using the backported `dataclasses` for new features, only to support Python 3.6.

* Our position as a much-depended on package requires us to be stable, taking on a dependency from PyPI w/ the CalVer versioning scheme is riskier for us compared to using stdlib imo, not sure if anything that attrs brings to the table that we'd use is worth that risk?

* Thankfully a lot of the re-packagers have `python3` to mean 3.7 or later nowadays, so the decision should already be made for them?